src/librustc_mir/transform/instcombine.rs

```diff
@@ -67,28 +77,485 @@ impl<'tcx> MutVisitor<'tcx> for InstCombineVisitor<'tcx> {
+                        ecx.report(&mut err, false, Some(span));
+                        None
+                    },
+                }
+            },
+        }
+    }
+}
+
+fn type_size_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                          param_env: ty::ParamEnv<'tcx>,
+                          ty: ty::Ty<'tcx>) -> Option<u64> {
+    use rustc::ty::layout::LayoutOf;
+    (tcx, param_env).layout_of(ty).ok().map(|layout| layout.size.bytes())
+}
+
+struct ConstPropVisitor {
+    local: Local,
+    can_const_prop: bool,
+    // false at the beginning, once set, there are not allowed to be any more assignments
+    found_assignment: bool,
+}
+
+impl ConstPropVisitor {
+    /// returns true if `local` can be propagated
+    fn check<'tcx>(local: Local, mir: &Mir<'tcx>) -> bool {
+        let mut cpv = ConstPropVisitor {
+            local,
+            can_const_prop: true,
+            found_assignment: false,
+        };
+        cpv.visit_mir(mir);
+        cpv.can_const_prop
+    }
+
+    fn is_our_local(&mut self, mut place: &Place) -> bool {
+        while let Place::Projection(ref proj) = place {
+            place = &proj.base;
+        }
+        if let Place::Local(local) = *place {
+            local == self.local
+        } else {
+            false
+        }
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for ConstPropVisitor {
+    fn visit_statement(
+        &mut self,
+        block: BasicBlock,
+        statement: &Statement<'tcx>,
+        location: Location,
+    ) {
+        self.super_statement(block, statement, location);
+        match statement.kind {
+            StatementKind::SetDiscriminant { ref place, .. } |
+            StatementKind::Assign(ref place, _) => {
+                if self.is_our_local(place) {
+                    if self.found_assignment {
+                        self.can_const_prop = false;
+                    } else {
+                        self.found_assignment = true
+                    }
+                }
+            },
+            StatementKind::InlineAsm { ref outputs, .. } => {
+                for place in outputs {
+                    if self.is_our_local(place) {
+                        if self.found_assignment {
+                            self.can_const_prop = false;
+                        } else {
+                            self.found_assignment = true
+                        }
+                        return;
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+    fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
+        self.super_rvalue(rvalue, location);
+        if let Rvalue::Ref(_, _, ref place) = *rvalue {
+            if self.is_our_local(place) {
+                self.can_const_prop = false;
+            }
+        }
+    }
 }
 
 impl<'b, 'a, 'tcx> Visitor<'tcx> for OptimizationFinder<'b, 'a, 'tcx> {
+    // override to visit basic blocks in execution order
```

Comment:
Did you reinvent/reimplement RPO? *Needing* a specific order is usually a pretty bad sign, it means correctness is much harder to understand, or it's not correct in the first place.

Reply:
The order is just important for anything to happen at all, not for correctness. You only get const prop for `_1 + _2` if `_1 = 42` and `_2 = 55` have already been visited.

Reply:
You can do it order-independent by detecting an unique initializer (as per https://github.com/rust-lang/rust/pull/46882#discussion_r163838334). That's how rvalue promotion works, for example.